homeassistant/components/smappee/config_flow.py

```diff
@@ -21,3 +28,123 @@ class SmappeeFlowHandler(
+        if user_input is None and not prepare:
+            if source == SOURCE_ZEROCONF:
+                return self._show_zeroconf_confirm_dialog()
+
+            # Show environment form with CLOUD or LOCAL option
+            return self._show_setup_form(step="environment")
+
+        # Environment chosen, request additional host information for LOCAL or OAuth2 flow for CLOUD
+        if user_input is not None and "environment" in user_input:
+            if user_input["environment"] == ENV_LOCAL:
+                self.context.update({"environment": ENV_LOCAL})
+                return self._show_setup_form(step="host")
+
+            # Use configuration.yaml CLOUD setup
+            self.context.update({"environment": ENV_CLOUD})
+            return await self.async_step_pick_implementation()
+
+        if source == SOURCE_ZEROCONF:
+            user_input[CONF_IP_ADDRESS] = self.context.get(CONF_IP_ADDRESS)
+            user_input[CONF_SERIALNUMBER] = self.context.get(CONF_SERIALNUMBER)
+        else:
+            user_input[CONF_IP_ADDRESS] = user_input["host"]
+
+        if user_input.get(CONF_IP_ADDRESS) is not None or not prepare:
+            smappee_api = api.api.SmappeeLocalApi(ip=user_input[CONF_IP_ADDRESS])
+            try:
+                await self.hass.async_add_executor_job(smappee_api.logon)
+
+                # In a LOCAL setup we still need to resolve the host to serialnumber
+                if not source == SOURCE_ZEROCONF:
+                    advanced_config = await self.hass.async_add_executor_job(
+                        smappee_api.load_advanced_config
+                    )
+                    serialnumber = None
+                    for config_item in advanced_config:
+                        if config_item["key"] == "mdnsHostName":
+                            serialnumber = config_item["value"]
+
+                    if serialnumber is None or not serialnumber.startswith("Smappee1"):
+                        # We currently only support Energy and Solar models (legacy)
+                        return self.async_abort(reason="invalid_mdns")
+                    user_input[CONF_SERIALNUMBER] = serialnumber.replace("Smappee", "")
+            except (socket.gaierror, socket.timeout, ConnectTimeout, ConnectionError):
+                return self.async_abort(reason="connection_error")
+
+        # Check if already configured
+        await self.async_set_unique_id(f"Smappee{user_input[CONF_SERIALNUMBER]}")
+        self._abort_if_unique_id_configured()
+
+        if prepare:
+            return await self.async_step_zeroconf_confirm()
+
+        return self.async_create_entry(
+            title=f"Smappee{user_input[CONF_SERIALNUMBER]}",
+            data={
+                CONF_IP_ADDRESS: user_input[CONF_IP_ADDRESS],
+                CONF_SERIALNUMBER: user_input[CONF_SERIALNUMBER],
+            },
+        )
+
+    def _show_setup_form(self, step, errors=None):
```

Comment:
Please remove the form methods and inline the show form in each step that should show a form.